test/parser_test.js

```diff
@@ -165,5 +210,450 @@ describe("Parser", () => {
         );
       })
     })
+
+    describe('invalid chars at eof', () => {
+      it('shows in the parse tree', () => {
+        parser.setLanguage(JSON);
+        const tree = parser.parse('\xdf');
+        assert.equal(tree.rootNode.toString(), "(document (ERROR (UNEXPECTED 223)))");
+      });
+    });
+
+    describe('unexpected null characters', () => {
+      it('has a null character in the source', () => {
+        parser.setLanguage(JavaScript);
+        const tree = parser.parse('var \0 something;')
+        assert.equal(
+          tree.rootNode.toString(),
+          "(program (variable_declaration (ERROR (UNEXPECTED '\\0')) (variable_declarator name: (identifier))))"
+        );
+      });
+    });
+
+    it("parses an empty file with a reused tree", () => {
+      parser.setLanguage(Rust);
+
+      let tree = parser.parse("");
+      parser.parse("", tree);
+
+      tree = parser.parse("\n ");
+      parser.parse("\n ", tree);
+    });
+
+    describe('parsing with a timeout', () => {
+      it('stops after a certain number of microseconds', () => {
+        parser.setLanguage(JSON);
+
+        // dont use Date.now(), we need microseconds
+        let startTime = performance.now() * 1000;
+        parser.setTimeoutMicros(1000);
+        let tree = parser.parse((offset, _) => offset === 0 ? " [" : ",0");
+        assert.equal(tree, null);
+        assert.isBelow(performance.now() * 1000 - startTime, 2000);
```

Comment:
`performance` is not available as a global in Node.js 14 and needs to be imported from `perf_hooks`.

Reply:
node 14 is eol, we should probably drop it